File: src/backend/app/services/s3_service.py
import boto3
from botocore.exceptions import ClientError
import json
import os
import logging

logging.basicConfig(
    level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s"
)
logger = logging.getLogger(__name__)

s3_client = boto3.client("s3")

BUCKET_NAME = os.getenv("S3_LLM_LOGS_BUCKET")
BUCKET_DIRECTORY = os.getenv("S3_LLM_LOGS_DIRECTORY")

logger.info(f"S3 Bucket Name: {BUCKET_NAME}")
logger.info(f"S3 Bucket Directory: {BUCKET_DIRECTORY}")


async def fetch_s3_file_history(batch: int, start_after: str = None):
    logger.info(
        f"Fetching S3 file history with batch size: {batch} and start_after: {start_after}"
    )

    paginator = s3_client.get_paginator("list_objects_v2")
    operation_params = {
        "Bucket": BUCKET_NAME,
        "MaxKeys": batch,
        "Prefix": BUCKET_DIRECTORY,
    }
    if start_after:
        operation_params["StartAfter"] = start_after

    try:
        page_iterator = paginator.paginate(**operation_params)
        files = []

        for page in page_iterator:
            logger.info(f"Received page with {len(page.get('Contents', []))} objects.")
            for obj in page.get("Contents", []):
                if not obj["Key"].endswith("/"):
                    files.append(
                        {
                            "name": obj["Key"],
                            "last_modified": obj["LastModified"].strftime(
                                "%Y-%m-%d %H:%M:%S"
                            ),
                        }
                    )
            break  # Only fetch one page at a time

        logger.info(f"Successfully fetched {len(files)} files.")
        return files

    except ClientError as e:
        logger.error(
            f"AWS ClientError fetching file history: {e.response['Error']['Message']}"
        )
        raise Exception(f"Error fetching S3 file history: {str(e)}")
    except Exception as e:
        logger.error(f"Unexpected error fetching file history: {str(e)}")
        raise Exception(f"Error fetching S3 file history: {str(e)}")


async def get_s3_file_content(file_name: str):
    logger.info(f"Attempting to fetch file content for key: {file_name}")

    try:
        response = s3_client.get_object(Bucket=BUCKET_NAME, Key=file_name)
        file_content = response["Body"].read().decode("utf-8")
        json_content = json.loads(file_content)

        logger.info(f"Successfully fetched and parsed content for file: {file_name}")
        return json_content

    except ClientError as e:
        logger.error(
            f"AWS ClientError fetching file content: {e.response['Error']['Message']}"
        )
        raise Exception(f"Error fetching S3 file content: {str(e)}")
    except Exception as e:
        logger.error(f"Unexpected error fetching file content: {str(e)}")
        raise Exception(f"Error fetching S3 file content: {str(e)}")

File: containers/backend/src/s3Service.ts
import { GetObjectCommand, S3Client, paginateListObjectsV2 } from '@aws-sdk/client-s3';

import type { Readable } from 'node:stream';

export const s3Client = new S3Client({});

const BUCKET_NAME = process.env.S3_LLM_LOGS_BUCKET;
const BUCKET_DIRECTORY = process.env.S3_LLM_LOGS_DIRECTORY;
const BUCKET_ACCOUNT_ID = process.env.S3_LLM_LOGS_BUCKET_ACCOUNT_ID;

console.info(`S3 Bucket Name: ${BUCKET_NAME}`);
console.info(`S3 Bucket Directory: ${BUCKET_DIRECTORY}`);
console.info(`S3 Bucket Account ID: ${BUCKET_ACCOUNT_ID}`);

export interface FileEntry {
  name: string;
  last_modified: string;
}

/** Formats a Date as YYYY-MM-DD HH:MM:SS (UTC) to match Python's strftime format. */
function formatDatetime(d: Date): string {
  return d.toISOString().replace('T', ' ').slice(0, 19);
}

export async function fetchS3FileHistory(): Promise<FileEntry[]> {
  const params = {
    Bucket: BUCKET_NAME,
    Prefix: BUCKET_DIRECTORY,
    ExpectedBucketOwner: BUCKET_ACCOUNT_ID,
  };

  try {
    const files: Array<{ name: string; lastModified: Date }> = [];

    for await (const page of paginateListObjectsV2({ client: s3Client }, params)) {
      const contents = page.Contents ?? [];
      console.info(`Received page with ${contents.length} objects.`);

      for (const obj of contents) {
        if (obj.Key && !obj.Key.endsWith('/') && obj.LastModified) {
          files.push({ name: obj.Key, lastModified: obj.LastModified });
        }
      }
    }

    files.sort((a, b) => b.lastModified.getTime() - a.lastModified.getTime());
    console.info(`Successfully fetched and sorted ${files.length} files.`);

    return files.map((f) => ({
      name: f.name,
      last_modified: formatDatetime(f.lastModified),
    }));
  } catch (error) {
    const msg = error instanceof Error ? error.message : String(error);
    console.error(`Error fetching S3 file history: ${msg}`);
    throw new Error(`Error fetching S3 file history: ${msg}`);
  }
}

export async function getS3FileContent(fileName: string): Promise<unknown> {
  console.info(`Attempting to fetch file content for key: ${fileName}`);

  try {
    const response = await s3Client.send(
      new GetObjectCommand({
        Bucket: BUCKET_NAME,
        Key: fileName,
        ExpectedBucketOwner: BUCKET_ACCOUNT_ID,
      }),
    );

    // AWS SDK v3 Node.js streams support transformToString() directly
    const text = await (response.Body as Readable & { transformToString(): Promise<string> }).transformToString();
    const parsed = JSON.parse(text) as unknown;

    console.info(`Successfully fetched and parsed content for file: ${fileName}`);
    return parsed;
  } catch (error) {
    const msg = error instanceof Error ? error.message : String(error);
    console.error(`Error fetching S3 file content: ${msg}`);
    throw new Error(`Error fetching S3 file content: ${msg}`);
  }
}
